### PR TITLE
golang: Update to 1.15.2, update/enhance Makefiles

### PR DESCRIPTION
**Pull request by jefferyto**
Maintainer: me
Compile tested: x86-64/armvirt-32/armvirt-64/malta-be, 2020-10-01 snapshot sdk
Run tested: x86-64/armvirt-32/armvirt-64/malta-be, 2020-10-01 snapshot

Description:
This updates the Go compiler to 1.15.2 and includes a number of updates/enhancements, most notably:
* Caches are now shared between packages/builds:
  * The module cache is moved into the download directory
  * The build cache is moved into the top-level directory (similar to ccache)
* The addition of golang-host-build.mk enables host builds of tools written in Go
* `GoPackage/Package/Install/Bin` can be told to install binaries to somewhere else (inside the package) other than `/usr/bin` by setting `GO_PKG_INSTALL_BIN_PATH` in the package Makefile
* Go program packages can be compiled with the [Spectre mitigations](https://github.com/golang/go/wiki/Spectre) added in Go 1.15 (currently only available for x86-64)

### CODE REVIEW

**Comment by jefferyto** [comment]
Can the code lint check be told to always ignore SC2039 (`In POSIX sh, 'local' is undefined.`)? `local` is a very useful construct and supported by bash/dash/busybox sh.

**Comment by neheb** [comment]
One of my suggestions was to fix it by passing -s dash.

Also, https://github.com/koalaman/shellcheck/wiki/SC2039 says it's deprecated. I assume new versions don't throw this warning.

**Comment by jefferyto** [comment]
I've added `disable` directives in the necessary places - the linter now passes.

File: lang/golang/golang/Makefile [hunk 1]
```diff
@@ -247,40 +247,38 @@ endef
...
 endif
 
 $(eval $(call GoCompiler/AddProfile,Package,$(PKG_BUILD_DIR),$(PKG_GO_PREFIX),$(PKG_GO_VERSION_ID),$(GO_OS_ARCH),$(PKG_GO_INSTALL_SUFFIX)))
 
 PKG_GO_ZBOOTSTRAP_MODS:= \
-	s/defaultGO386 = `[^`]*`/defaultGO386 = `$(if $(GO_386),$(GO_386),387)`/; \
-	s/defaultGOARM = `[^`]*`/defaultGOARM = `$(if $(GO_ARM),$(GO_ARM),5)`/; \
-	s/defaultGOMIPS = `[^`]*`/defaultGOMIPS = `$(if $(GO_MIPS),$(GO_MIPS),hardfloat)`/; \
-	s/defaultGOMIPS64 = `[^`]*`/defaultGOMIPS64 = `$(if $(GO_MIPS64),$(GO_MIPS64),hardfloat)`/; \
+	s/defaultGO386 = `[^`]*`/defaultGO386 = `$(or $(GO_386),387)`/; \
```
Comment (neheb) [inline lang/golang/golang/Makefile:261]: very interesting... I've never seen this before.

**Comment by neheb** [comment]
> I've added `disable` directives in the necessary places - the linter now passes.

https://github.com/koalaman/shellcheck/commit/cfd68ee0c2ebfd0ab08a1d4bf628162b454dc207 is not in a release version unfortunately. that way we can disable the bogus ones.

**Comment by neheb** [comment]
Now this is failing on all PRs...

```
 2020-10-07 10:10:35 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:35 [INFO  ]   ------ Files modified in the commit(s): ------
2020-10-07 10:10:35 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:35 [INFO  ]   File:[lang/golang/golang-compiler.mk], File_type:[mk], Base_file:[golang-compiler.mk]
2020-10-07 10:10:35 [WARN  ]   Failed to get filetype for:[lang/golang/golang-compiler.mk]!
2020-10-07 10:10:35 [INFO  ]   File:[lang/golang/golang-package.mk], File_type:[mk], Base_file:[golang-package.mk]
2020-10-07 10:10:36 [WARN  ]   Failed to get filetype for:[lang/golang/golang-package.mk]!
2020-10-07 10:10:36 [INFO  ]   File:[lang/golang/golang-values.mk], File_type:[mk], Base_file:[golang-values.mk]
2020-10-07 10:10:36 [WARN  ]   Failed to get filetype for:[lang/golang/golang-values.mk]!
2020-10-07 10:10:36 [INFO  ]   File:[lang/golang/golang/Makefile], File_type:[lang/golang/golang/makefile], Base_file:[makefile]
2020-10-07 10:10:36 [WARN  ]   Failed to get filetype for:[lang/golang/golang/Makefile]!
2020-10-07 10:10:36 [INFO  ]   File:[lang/golang/golang/files/go-gcc-helper], File_type:[lang/golang/golang/files/go-gcc-helper], Base_file:[go-gcc-helper]
2020-10-07 10:10:36 [INFO  ]   File:[libs/libshout/Makefile], File_type:[libs/libshout/makefile], Base_file:[makefile]
2020-10-07 10:10:36 [WARN  ]   Failed to get filetype for:[libs/libshout/Makefile]!
2020-10-07 10:10:36 [INFO  ]   File:[libs/libshout/patches/130-usleep.patch], File_type:[patch], Base_file:[130-usleep.patch]
2020-10-07 10:10:36 [WARN  ]   Failed to get filetype for:[libs/libshout/patches/130-usleep.patch]!
2020-10-07 10:10:36 [INFO  ]   File:[utils/ctop/Makefile], File_type:[utils/ctop/makefile], Base_file:[makefile]
2020-10-07 10:10:36 [WARN  ]   Failed to get filetype for:[utils/ctop/Makefile]!
2020-10-07 10:10:36 [INFO  ]   File:[utils/docker-ce/Makefile], File_type:[utils/docker-ce/makefile], Base_file:[makefile]
2020-10-07 10:10:36 [WARN  ]   Failed to get filetype for:[utils/docker-ce/Makefile]!
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   Successfully gathered list of files...
2020-10-07 10:10:36 [WARN  ]   No Ansible base directory found at:[/github/workspace/ansible]
2020-10-07 10:10:36 [INFO  ]   
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   Linting [BASH] files...
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   ---------------------------
2020-10-07 10:10:36 [INFO  ]   File:[lang/golang/golang/files/go-gcc-helper]
2020-10-07 10:10:36 [ERROR ]   Found errors in [shellcheck] linter!
2020-10-07 10:10:36 [ERROR ]   [
In lang/golang/golang/files/go-gcc-helper line 4:
name=$(basename $0)
                ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
name=$(basename "$0")


In lang/golang/golang/files/go-gcc-helper line 21:
echo "$me: running $cmd $@"
                        ^-- SC2145: Argument mixes string and array. Use * or separate argument.

For more information:
  https://www.shellcheck.net/wiki/SC2145 -- Argument mixes string and array. ...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...]
2020-10-07 10:10:36 [ERROR ]   Linter CMD:[shellcheck --color --external-sources lang/golang/golang/files/go-gcc-helper]
2020-10-07 10:10:36 [INFO  ]   
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   Linting [BASH_EXEC] files...
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:36 [INFO  ]   ---------------------------
2020-10-07 10:10:36 [INFO  ]   File:[lang/golang/golang/files/go-gcc-helper]
2020-10-07 10:10:36 [WARN  ]   Warnings found in [bash-exec] linter!
2020-10-07 10:10:36 [WARN  ]   Error: File:[lang/golang/golang/files/go-gcc-helper] is not executable
2020-10-07 10:10:36 [WARN  ]   No Kubernetes directory found at:[/github/workspace/kubernetes]
2020-10-07 10:10:39 [WARN  ]   No .editorconfig found at:[/github/workspace/.editorconfig]
2020-10-07 10:10:39 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:39 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:39 [INFO  ]   Generated reports:
2020-10-07 10:10:39 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:39 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:39 [INFO  ]   Reports generated in folder /github/workspace/super-linter.report
2020-10-07 10:10:39 [WARN  ]   Report output folder (/github/workspace/super-linter.report) does NOT exist.
2020-10-07 10:10:39 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:39 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:39 [INFO  ]   The script has completed
2020-10-07 10:10:39 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:39 [INFO  ]   ----------------------------------------------
2020-10-07 10:10:39 [ERROR ]   ERRORS FOUND in BASH:[1]
2020-10-07 10:10:40 [INFO  ]   ERROR! Failed to call GitHub Status API!
2020-10-07 10:10:40 [INFO  ]   ERROR:[]
2020-10-07 10:10:40 [INFO  ]   ERROR! Failed to call GitHub Status API!
2020-10-07 10:10:40 [INFO  ]   ERROR:[]
2020-10-07 10:10:40 [FATAL ]   Exiting with errors found!
```

**Comment by jefferyto** [comment]
This is an issue with the linter - perhaps @aparcar can help to investigate.

**Pull request merged**